File: orchestrator/tests/test_canvas_file_format.py
#!/usr/bin/env python3
"""
Unit tests for ``orchestrator/canvas_file_format.py`` — WP-7.0.2.

Runs under stdlib ``unittest`` (no pytest)::

    /opt/homebrew/bin/python3 -m unittest discover -s ~/ora/orchestrator/tests -v
    # or
    /opt/homebrew/bin/python3 ~/ora/orchestrator/tests/test_canvas_file_format.py

Coverage maps directly to the WP-7.0.2 test brief:

    "Round-trip a canvas with 5 objects + 2 embedded images through both
     compressed and uncompressed paths; visual recovery exact; bytes
     equivalent after decompression."

Plus structural-validator coverage, gzip-magic detection, and the schema
file's existence + parsability check.
"""
from __future__ import annotations

import base64
import gzip
import io
import json
import os
import sys
import tempfile
import unittest
from pathlib import Path

HERE = Path(__file__).resolve().parent
ORCHESTRATOR = HERE.parent
sys.path.insert(0, str(ORCHESTRATOR))

from canvas_file_format import (  # noqa: E402
    DEFAULT_CANVAS_H,
    DEFAULT_CANVAS_W,
    FORMAT_ID,
    SCHEMA_PATH,
    SCHEMA_VERSION,
    is_compressed,
    new_canvas_state,
    read_bytes,
    read_path,
    validate,
    write_bytes,
    write_json_string,
    write_path,
)


# ── Fixtures ────────────────────────────────────────────────────────────────

# Two tiny but real-looking PNGs encoded as base64. The bytes are arbitrary —
# they only need to be distinct, non-empty, and survive a JSON round-trip.
_PNG_A_BYTES = bytes(range(256)) * 4          # 1024 bytes, deterministic
_PNG_B_BYTES = bytes(reversed(range(256))) * 3  # 768 bytes, deterministic

_PNG_A_B64 = base64.b64encode(_PNG_A_BYTES).decode("ascii")
_PNG_B_B64 = base64.b64encode(_PNG_B_BYTES).decode("ascii")


def _make_test_canvas() -> dict:
    """Build a canvas state with 5 objects (3 shapes + 2 images), plus a
    non-default view, exercising every required schema field."""
    state = new_canvas_state(
        title="Round-trip fixture",
        conversation_id="conv-12345",
        ora_version="ora-test",
        # Pin the timestamp so two writes of this fixture compare equal byte-for-byte.
        now="2026-04-29T12:00:00+00:00",
    )
    state["view"] = {"zoom": 1.5, "pan": {"x": -250.5, "y": 80}}
    state["objects"] = [
        # Three Konva-style shapes on user_input.
        {
            "id": "u-rect-0",
            "kind": "shape",
            "layer": "user_input",
            "konva_class": "Rect",
            "x": 100, "y": 200, "width": 80, "height": 40,
            "rotation": 0, "scale_x": 1, "scale_y": 1, "opacity": 1, "visible": True,
            "user_label": "Cause",
            "attrs": {"fill": "#aabbcc", "stroke": "#112233", "strokeWidth": 2},
        },
        {
            "id": "u-rect-1",
            "kind": "shape",
            "layer": "user_input",
            "konva_class": "Rect",
            "x": 300, "y": 200, "width": 80, "height": 40,
            "user_label": "Effect",
            "attrs": {"fill": "#ddeeff", "stroke": "#112233", "strokeWidth": 2},
        },
        {
            "id": "u-arr-0",
            "kind": "shape",
            "layer": "user_input",
            "konva_class": "Arrow",
            "user_label": "→",
            "attrs": {"points": [180, 220, 300, 220], "stroke": "#000", "strokeWidth": 2},
        },
        # Two embedded images on background.
        {
            "id": "img-bg-0",
            "kind": "image",
            "layer": "background",
            "x": 0, "y": 0, "width": 800, "height": 600,
            "image_data": {
                "mime_type": "image/png",
                "encoding": "base64",
                "data": _PNG_A_B64,
                "natural_width": 800,
                "natural_height": 600,
                "source": "upload",
            },
        },
        {
            "id": "img-overlay-0",
            "kind": "image",
            "layer": "annotation",
            "x": 50, "y": 50, "width": 200, "height": 150,
            "image_data": {
                "mime_type": "image/jpeg",
                "encoding": "base64",
                "data": _PNG_B_B64,
                "natural_width": 200,
                "natural_height": 150,
                "source": "generated:test-provider",
            },
        },
    ]
    return state


# ── Tests ───────────────────────────────────────────────────────────────────

class TestSchemaFileExists(unittest.TestCase):
    """The schema doc lives at ~/ora/config/schemas/canvas-state.schema.json
    per WP-7.0.2 deliverable. Confirm it's present and well-formed JSON."""

    def test_schema_file_present(self) -> None:
        self.assertTrue(SCHEMA_PATH.exists(), f"schema not found at {SCHEMA_PATH}")

    def test_schema_file_parses(self) -> None:
        doc = json.loads(SCHEMA_PATH.read_text())
        self.assertEqual(doc["$schema"], "https://json-schema.org/draft/2020-12/schema")
        self.assertEqual(doc["title"], "Ora Canvas State v0.1")
        self.assertEqual(doc["additionalProperties"], False)
        # Sanity-check the discriminator for objects.
        defs = doc.get("$defs", {}).get("object", {})
        self.assertEqual(defs.get("additionalProperties"), False)


class TestNewCanvasState(unittest.TestCase):
    def test_factory_defaults(self) -> None:
        s = new_canvas_state()
        self.assertEqual(s["schema_version"], SCHEMA_VERSION)
        self.assertEqual(s["format_id"], FORMAT_ID)
        self.assertEqual(s["metadata"]["canvas_size"]["width"], DEFAULT_CANVAS_W)
        self.assertEqual(s["metadata"]["canvas_size"]["height"], DEFAULT_CANVAS_H)
        self.assertEqual(s["view"]["zoom"], 1)
        self.assertEqual(s["view"]["pan"], {"x": 0, "y": 0})
        kinds = [L["kind"] for L in s["layers"]]
        self.assertEqual(kinds, ["background", "annotation", "user_input", "selection"])
        self.assertEqual(s["objects"], [])

    def test_factory_overrides(self) -> None:
        s = new_canvas_state(canvas_size={"width": 4096, "height": 4096}, title="t", conversation_id="c-1")
        self.assertEqual(s["metadata"]["canvas_size"], {"width": 4096, "height": 4096})
        self.assertEqual(s["metadata"]["title"], "t")
        self.assertEqual(s["metadata"]["conversation_id"], "c-1")


class TestRoundTrip(unittest.TestCase):
    """The WP-7.0.2 test criterion in three parts."""

    def test_compressed_round_trip_visual_exact(self) -> None:
        s = _make_test_canvas()
        blob = write_bytes(s, compressed=True)
        self.assertTrue(is_compressed(blob), "compressed output missing gzip magic header")
        recovered = read_bytes(blob)
        self.assertEqual(recovered, s, "compressed round-trip lost data")

    def test_uncompressed_round_trip_visual_exact(self) -> None:
        s = _make_test_canvas()
        blob = write_bytes(s, compressed=False)
        self.assertFalse(is_compressed(blob), "uncompressed output should not have gzip magic")
        recovered = read_bytes(blob)
        self.assertEqual(recovered, s, "uncompressed round-trip lost data")

    def test_bytes_equivalent_after_decompression(self) -> None:
        """Plan §11.7 contract: ‘Both round-trip identically.’ This is the
        load-bearing invariant — the compressed and uncompressed surfacing
        forms wrap the SAME bytes."""
        s = _make_test_canvas()
        compressed   = write_bytes(s, compressed=True)
        uncompressed = write_bytes(s, compressed=False)
        decompressed = gzip.decompress(compressed)
        self.assertEqual(
            decompressed, uncompressed,
            "decompressed gzip output diverges from raw JSON output",
        )

    def test_compressed_smaller_than_uncompressed(self) -> None:
        """Sanity check that gzip is doing useful work on a payload that
        contains base64-encoded image data (which compresses very well)."""
        s = _make_test_canvas()
        c = write_bytes(s, compressed=True)
        u = write_bytes(s, compressed=False)
        self.assertLess(len(c), len(u), "compressed should be smaller than uncompressed for this fixture")

    def test_image_payloads_recovered_byte_exact(self) -> None:
        s = _make_test_canvas()
        recovered = read_bytes(write_bytes(s, compressed=True))
        for orig_obj, rec_obj in zip(s["objects"], recovered["objects"]):
            if orig_obj.get("kind") == "image":
                self.assertEqual(orig_obj["image_data"]["data"], rec_obj["image_data"]["data"])
                # Decode and confirm bytes match the source fixture.
                if orig_obj["id"] == "img-bg-0":
                    self.assertEqual(base64.b64decode(rec_obj["image_data"]["data"]), _PNG_A_BYTES)
                if orig_obj["id"] == "img-overlay-0":
                    self.assertEqual(base64.b64decode(rec_obj["image_data"]["data"]), _PNG_B_BYTES)

    def test_canonicalisation_stable_across_writes(self) -> None:
        """Two writes of the same canvas should produce identical bytes —
        otherwise diff-based persistence (autosave, vault export) is
        unreliable."""
        s = _make_test_canvas()
        b1 = write_bytes(s, compressed=False)
        b2 = write_bytes(s, compressed=False)
        self.assertEqual(b1, b2)
        gz1 = write_bytes(s, compressed=True)
        gz2 = write_bytes(s, compressed=True)
        self.assertEqual(gz1, gz2)

    def test_canonicalisation_orders_keys(self) -> None:
        """Permuting key insertion order must not change the output bytes."""
        a = {"format_id": "ora-canvas", "schema_version": "0.1.0", "metadata": {"canvas_size": {"width": 10, "height": 10}}, "view": {"zoom": 1, "pan": {"x": 0, "y": 0}}, "layers": [{"id": "background", "kind": "background"}], "objects": []}
        b = {"layers": [{"kind": "background", "id": "background"}], "objects": [], "view": {"pan": {"y": 0, "x": 0}, "zoom": 1}, "metadata": {"canvas_size": {"height": 10, "width": 10}}, "schema_version": "0.1.0", "format_id": "ora-canvas"}
        self.assertEqual(write_bytes(a, compressed=False), write_bytes(b, compressed=False))


class TestPathHelpers(unittest.TestCase):
    def test_write_and_read_compressed_path(self) -> None:
        s = _make_test_canvas()
        with tempfile.TemporaryDirectory() as tmp:
            p = Path(tmp) / "session.ora-canvas"
            write_path(s, p)
            self.assertTrue(p.exists())
            self.assertTrue(is_compressed(p.read_bytes()))
            self.assertEqual(read_path(p), s)

    def test_write_and_read_uncompressed_path(self) -> None:
        s = _make_test_canvas()
        with tempfile.TemporaryDirectory() as tmp:
            p = Path(tmp) / "session.ora-canvas.json"
            write_path(s, p)
            self.assertTrue(p.exists())
            self.assertFalse(is_compressed(p.read_bytes()))
            self.assertEqual(read_path(p), s)
            # Auto-detection: gunzip magic is absent so read_path treats
            # this as raw JSON regardless of extension.

    def test_unknown_extension_rejected(self) -> None:
        s = _make_test_canvas()
        with tempfile.TemporaryDirectory() as tmp:
            with self.assertRaises(ValueError):
                write_path(s, Path(tmp) / "session.canvas")


class TestReadEdgeCases(unittest.TestCase):
    def test_format_id_mismatch_raises(self) -> None:
        bad = json.dumps({"format_id": "not-ora-canvas"}).encode("utf-8")
        with self.assertRaises(ValueError):
            read_bytes(bad)

    def test_invalid_json_raises(self) -> None:
        with self.assertRaises(ValueError):
            read_bytes(b"not json at all")

    def test_top_level_must_be_object(self) -> None:
        with self.assertRaises(ValueError):
            read_bytes(json.dumps([1, 2, 3]).encode("utf-8"))

    def test_auto_detect_compressed(self) -> None:
        s = _make_test_canvas()
        gz = gzip.compress(json.dumps(s).encode("utf-8"), mtime=0)
        # Force-feed unconventional compression source — read_bytes should
        # still gunzip then parse.
        recovered = read_bytes(gz)
        # JSON passes through json.loads, so equality is via Python value
        # comparison (key order does not matter).
        self.assertEqual(recovered, s)


class TestValidator(unittest.TestCase):
    def test_factory_state_validates(self) -> None:
        ok, errs = validate(new_canvas_state())
        self.assertTrue(ok, f"empty factory state failed validation: {errs}")

    def test_round_trip_state_validates(self) -> None:
        ok, errs = validate(_make_test_canvas())
        self.assertTrue(ok, f"round-trip fixture failed validation: {errs}")

    def test_missing_format_id_fails(self) -> None:
        bad = _make_test_canvas()
        bad.pop("format_id")
        ok, errs = validate(bad)
        self.assertFalse(ok)

    def test_image_without_image_data_fails(self) -> None:
        bad = _make_test_canvas()
        bad["objects"].append({"id": "broken", "kind": "image", "layer": "background"})
        ok, errs = validate(bad)
        self.assertFalse(ok)
        self.assertTrue(any("image_data" in e["message"] or "image_data" in e["path"] for e in errs))

    def test_image_data_url_prefix_rejected(self) -> None:
        bad = _make_test_canvas()
        bad["objects"][3]["image_data"]["data"] = "data:image/png;base64," + _PNG_A_B64
        ok, errs = validate(bad)
        self.assertFalse(ok)

    def test_unknown_top_level_property_fails(self) -> None:
        bad = _make_test_canvas()
        bad["surprise"] = True
        ok, errs = validate(bad)
        self.assertFalse(ok)

    def test_duplicate_object_id_fails_structural(self) -> None:
        bad = _make_test_canvas()
        bad["objects"].append({"id": "u-rect-0", "kind": "shape", "layer": "user_input"})
        ok, errs = validate(bad, use_jsonschema=False)
        self.assertFalse(ok)
        self.assertTrue(any("duplicate" in e["message"].lower() for e in errs))

    def test_invalid_layer_kind_fails(self) -> None:
        bad = _make_test_canvas()
        bad["layers"][0]["kind"] = "garbage"
        ok, errs = validate(bad)
        self.assertFalse(ok)


class TestGroupNesting(unittest.TestCase):
    def test_group_with_children_round_trips(self) -> None:
        s = new_canvas_state()
        s["objects"] = [
            {
                "id": "g-1",
                "kind": "group",
                "layer": "user_input",
                "children": [
                    {"id": "g-1.r-1", "kind": "shape", "layer": "user_input", "konva_class": "Rect", "attrs": {"fill": "red"}},
                    {"id": "g-1.r-2", "kind": "shape", "layer": "user_input", "konva_class": "Rect", "attrs": {"fill": "blue"}},
                ],
            }
        ]
        ok, errs = validate(s)
        self.assertTrue(ok, f"group fixture failed validation: {errs}")
        self.assertEqual(read_bytes(write_bytes(s)), s)

    def test_group_without_children_fails(self) -> None:
        s = new_canvas_state()
        s["objects"] = [{"id": "g-bad", "kind": "group", "layer": "user_input"}]
        ok, errs = validate(s)
        self.assertFalse(ok)


class TestWriteJsonString(unittest.TestCase):
    def test_pretty_print(self) -> None:
        s = _make_test_canvas()
        out = write_json_string(s, indent=2)
        self.assertIn("\n", out)
        # Re-parse and confirm equivalence.
        self.assertEqual(json.loads(out), s)

    def test_compact_default(self) -> None:
        s = _make_test_canvas()
        out = write_json_string(s)
        self.assertNotIn("\n", out)
        self.assertEqual(json.loads(out), s)


if __name__ == "__main__":
    unittest.main(verbosity=2)

File: server/static/ora-visual-compiler/tests/cases/test-canvas-file-format.js
/**
 * tests/cases/test-canvas-file-format.js — WP-7.0.2 regression suite.
 *
 * Exports { label, run(ctx, record) } per the run.js case-file convention.
 *
 * Coverage of the canvas-state file-format reader/writer
 * (server/static/canvas-file-format.js):
 *
 *   1. Namespace exists on window.OraCanvasFileFormat with the documented
 *      surface (read, write, validate, isCompressed, newCanvasState,
 *      readJsonBytes, writeJsonBytes, writeJsonString).
 *   2. newCanvasState() emits a valid skeleton (4 layers, default view).
 *   3. Compressed round-trip: 5 objects + 2 embedded base64 images survive
 *      write→read with deepEqual. (Direct test-brief coverage for WP-7.0.2.)
 *   4. Uncompressed round-trip: same fixture, raw JSON path.
 *   5. Bytes-equivalent-after-decompression: gunzip(compressed) === raw JSON.
 *   6. Compressed output is smaller than uncompressed (gzip is doing work).
 *   7. Image base64 payloads survive byte-exact through the round-trip.
 *   8. Canonicalisation: two writes of the same canvas produce identical
 *      bytes; permuted key order yields identical bytes.
 *   9. read() rejects non-canvas-state JSON (format_id mismatch).
 *  10. read() rejects malformed JSON.
 *  11. validate() accepts the round-trip fixture and rejects an image
 *      object lacking image_data.
 *  12. isCompressed() detects gzip magic bytes correctly.
 */

'use strict';

const path = require('path');
const fs   = require('fs');
const zlib = require('zlib');

// ── Test fixture ────────────────────────────────────────────────────────────

// Two distinct deterministic byte buffers, base64-encoded. Mirrors the
// Python suite's _PNG_A_BYTES / _PNG_B_BYTES so a future cross-language
// fixture share is straightforward.
function _makeBytesA() {
  var b = new Uint8Array(1024);
  for (var i = 0; i < b.length; i++) b[i] = i % 256;
  return b;
}
function _makeBytesB() {
  var b = new Uint8Array(768);
  for (var i = 0; i < b.length; i++) b[i] = (255 - (i % 256)) & 0xFF;
  return b;
}

function _toBase64(uint8) {
  // Node Buffer is the simplest path inside the jsdom harness.
  return Buffer.from(uint8).toString('base64');
}

function _fromBase64(b64) {
  return new Uint8Array(Buffer.from(b64, 'base64'));
}

function _u8Equal(a, b) {
  if (a.length !== b.length) return false;
  for (var i = 0; i < a.length; i++) if (a[i] !== b[i]) return false;
  return true;
}

function _deepEqual(a, b) {
  if (a === b) return true;
  if (a === null || b === null) return a === b;
  if (typeof a !== 'object' || typeof b !== 'object') return a === b;
  if (Array.isArray(a) !== Array.isArray(b)) return false;
  if (Array.isArray(a)) {
    if (a.length !== b.length) return false;
    for (var i = 0; i < a.length; i++) if (!_deepEqual(a[i], b[i])) return false;
    return true;
  }
  var ka = Object.keys(a).sort();
  var kb = Object.keys(b).sort();
  if (ka.length !== kb.length) return false;
  for (var k = 0; k < ka.length; k++) {
    if (ka[k] !== kb[k]) return false;
    if (!_deepEqual(a[ka[k]], b[kb[k]])) return false;
  }
  return true;
}

function makeCanvasFixture() {
  var pngA = _toBase64(_makeBytesA());
  var pngB = _toBase64(_makeBytesB());
  return {
    schema_version: '0.1.0',
    format_id:      'ora-canvas',
    metadata: {
      title:           'Round-trip fixture',
      conversation_id: 'conv-12345',
      ora_version:     'ora-test',
      created_at:      '2026-04-29T12:00:00.000Z',
      modified_at:     '2026-04-29T12:00:00.000Z',
      canvas_size:     { width: 10000, height: 10000 }
    },
    view: { zoom: 1.5, pan: { x: -250.5, y: 80 } },
    layers: [
      { id: 'background',  kind: 'background',  visible: true, locked: false, opacity: 1 },
      { id: 'annotation',  kind: 'annotation',  visible: true, locked: false, opacity: 1 },
      { id: 'user_input',  kind: 'user_input',  visible: true, locked: false, opacity: 1 },
      { id: 'selection',   kind: 'selection',   visible: true, locked: false, opacity: 1 }
    ],
    objects: [
      { id: 'u-rect-0', kind: 'shape', layer: 'user_input', konva_class: 'Rect',
        x: 100, y: 200, width: 80, height: 40, rotation: 0, scale_x: 1, scale_y: 1,
        opacity: 1, visible: true, user_label: 'Cause',
        attrs: { fill: '#aabbcc', stroke: '#112233', strokeWidth: 2 } },
      { id: 'u-rect-1', kind: 'shape', layer: 'user_input', konva_class: 'Rect',
        x: 300, y: 200, width: 80, height: 40, user_label: 'Effect',
        attrs: { fill: '#ddeeff', stroke: '#112233', strokeWidth: 2 } },
      { id: 'u-arr-0',  kind: 'shape', layer: 'user_input', konva_class: 'Arrow',
        user_label: '→',
        attrs: { points: [180, 220, 300, 220], stroke: '#000', strokeWidth: 2 } },
      { id: 'img-bg-0', kind: 'image', layer: 'background',
        x: 0, y: 0, width: 800, height: 600,
        image_data: { mime_type: 'image/png', encoding: 'base64', data: pngA,
                      natural_width: 800, natural_height: 600, source: 'upload' } },
      { id: 'img-overlay-0', kind: 'image', layer: 'annotation',
        x: 50, y: 50, width: 200, height: 150,
        image_data: { mime_type: 'image/jpeg', encoding: 'base64', data: pngB,
                      natural_width: 200, natural_height: 150,
                      source: 'generated:test-provider' } }
    ]
  };
}

// ── Suite ───────────────────────────────────────────────────────────────────

module.exports = {
  label: 'Canvas file format (WP-7.0.2) — .ora-canvas + .ora-canvas.json round-trip',
  run: async function run(ctx, record) {
    var win = ctx.win;

    // The module file lives one level above the compiler dir under
    // server/static/. Load it inside the jsdom context so it registers on
    // the window namespace.
    var modulePath = path.resolve(__dirname, '..', '..', '..', 'canvas-file-format.js');
    if (!fs.existsSync(modulePath)) {
      record('canvas-file-format: module file present', false, 'not found at ' + modulePath);
      return;
    }
    var src = fs.readFileSync(modulePath, 'utf-8');
    // win.eval lands the IIFE assignment on win.OraCanvasFileFormat.
    win.eval(src);

    if (typeof win.OraCanvasFileFormat === 'undefined') {
      record('canvas-file-format: namespace present', false, 'window.OraCanvasFileFormat undefined');
      return;
    }
    var F = win.OraCanvasFileFormat;

    // jsdom's window does NOT expose CompressionStream; inject Node's zlib
    // as the gzip implementation so the streams-API path doesn't fire.
    F._setGzipImpl({
      deflate: function (bytes) {
        // Mirror Python's mtime=0 + level=6 so the byte-equivalence
        // invariant holds across both languages where that matters.
        var buf = Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength);
        var out = zlib.gzipSync(buf, { level: 6 });
        return new Uint8Array(out.buffer, out.byteOffset, out.byteLength);
      },
      inflate: function (bytes) {
        var buf = Buffer.from(bytes.buffer, bytes.byteOffset, bytes.byteLength);
        var out = zlib.gunzipSync(buf);
        return new Uint8Array(out.buffer, out.byteOffset, out.byteLength);
      }
    });

    // 1. Namespace surface.
    record('canvas-file-format #1: namespace surface',
      typeof F === 'object' &&
      typeof F.read === 'function' &&
      typeof F.write === 'function' &&
      typeof F.validate === 'function' &&
      typeof F.isCompressed === 'function' &&
      typeof F.newCanvasState === 'function' &&
      typeof F.readJsonBytes === 'function' &&
      typeof F.writeJsonBytes === 'function' &&
      typeof F.writeJsonString === 'function' &&
      F.SCHEMA_VERSION === '0.1.0' &&
      F.FORMAT_ID === 'ora-canvas',
      'keys=' + Object.keys(F).join(','));

    // 2. newCanvasState skeleton.
    try {
      var skel = F.newCanvasState();
      var v = F.validate(skel);
      record('canvas-file-format #2: newCanvasState validates',
        v.valid && skel.layers.length === 4 && skel.view.zoom === 1,
        v.valid ? '' : JSON.stringify(v.errors));
    } catch (e) {
      record('canvas-file-format #2', false, 'threw: ' + (e.stack || e.message));
    }

    // 3. Compressed round-trip — 5 objects + 2 embedded images, deepEqual.
    var fixture = makeCanvasFixture();
    var compressedBytes;
    try {
      compressedBytes = await F.write(fixture, { compressed: true });
      var hasMagic = compressedBytes.length >= 2 && compressedBytes[0] === 0x1F && compressedBytes[1] === 0x8B;
      var recovered = await F.read(compressedBytes);
      record('canvas-file-format #3: compressed round-trip (5 objects + 2 images)',
        hasMagic && _deepEqual(recovered, fixture),
        hasMagic ? '' : 'gzip magic missing');
    } catch (e) {
      record('canvas-file-format #3', false, 'threw: ' + (e.stack || e.message));
    }

    // 4. Uncompressed round-trip.
    var uncompressedBytes;
    try {
      uncompressedBytes = await F.write(fixture, { compressed: false });
      var noMagic = !(uncompressedBytes[0] === 0x1F && uncompressedBytes[1] === 0x8B);
      var recovered = await F.read(uncompressedBytes);
      record('canvas-file-format #4: uncompressed round-trip',
        noMagic && _deepEqual(recovered, fixture),
        noMagic ? '' : 'unexpected gzip magic on uncompressed output');
    } catch (e) {
      record('canvas-file-format #4', false, 'threw: ' + (e.stack || e.message));
    }

    // 5. Bytes-equivalent-after-decompression.
    try {
      var c = await F.write(fixture, { compressed: true });
      var u = await F.write(fixture, { compressed: false });
      var dec = zlib.gunzipSync(Buffer.from(c.buffer, c.byteOffset, c.byteLength));
      var decU8 = new Uint8Array(dec.buffer, dec.byteOffset, dec.byteLength);
      record('canvas-file-format #5: gunzip(compressed) === uncompressed JSON bytes',
        _u8Equal(decU8, u),
        _u8Equal(decU8, u) ? '' :
          'decompressed=' + decU8.length + 'B, raw=' + u.length + 'B');
    } catch (e) {
      record('canvas-file-format #5', false, 'threw: ' + (e.stack || e.message));
    }

    // 6. Compression effectiveness.
    try {
      var c = await F.write(fixture, { compressed: true });
      var u = await F.write(fixture, { compressed: false });
      record('canvas-file-format #6: compressed smaller than uncompressed',
        c.length < u.length,
        'compressed=' + c.length + 'B, uncompressed=' + u.length + 'B');
    } catch (e) {
      record('canvas-file-format #6', false, 'threw: ' + (e.stack || e.message));
    }

    // 7. Image bytes survive byte-exact.
    try {
      var c = await F.write(fixture, { compressed: true });
      var rec = await F.read(c);
      var imgA = rec.objects.find(function (o) { return o.id === 'img-bg-0'; });
      var imgB = rec.objects.find(function (o) { return o.id === 'img-overlay-0'; });
      var aBytes = _fromBase64(imgA.image_data.data);
      var bBytes = _fromBase64(imgB.image_data.data);
      var aMatch = _u8Equal(aBytes, _makeBytesA());
      var bMatch = _u8Equal(bBytes, _makeBytesB());
      record('canvas-file-format #7: image base64 payloads recovered byte-exact',
        aMatch && bMatch,
        'A:' + (aMatch ? 'ok' : 'MISMATCH') + ' B:' + (bMatch ? 'ok' : 'MISMATCH'));
    } catch (e) {
      record('canvas-file-format #7', false, 'threw: ' + (e.stack || e.message));
    }

    // 8a. Canonicalisation stable across writes.
    try {
      var b1 = await F.write(fixture, { compressed: false });
      var b2 = await F.write(fixture, { compressed: false });
      record('canvas-file-format #8a: two writes of same canvas → identical bytes',
        _u8Equal(b1, b2),
        '');
    } catch (e) {
      record('canvas-file-format #8a', false, 'threw: ' + (e.stack || e.message));
    }

    // 8b. Canonicalisation orders keys.
    try {
      var a = { format_id: 'ora-canvas', schema_version: '0.1.0',
                metadata: { canvas_size: { width: 10, height: 10 } },
                view: { zoom: 1, pan: { x: 0, y: 0 } },
                layers: [{ id: 'background', kind: 'background' }],
                objects: [] };
      var b = { layers: [{ kind: 'background', id: 'background' }], objects: [],
                view: { pan: { y: 0, x: 0 }, zoom: 1 },
                metadata: { canvas_size: { height: 10, width: 10 } },
                schema_version: '0.1.0', format_id: 'ora-canvas' };
      var ba = await F.write(a, { compressed: false });
      var bb = await F.write(b, { compressed: false });
      record('canvas-file-format #8b: permuted key order yields identical bytes',
        _u8Equal(ba, bb),
        '');
    } catch (e) {
      record('canvas-file-format #8b', false, 'threw: ' + (e.stack || e.message));
    }

    // 9. format_id mismatch rejected.
    try {
      var bad = new TextEncoder().encode(JSON.stringify({ format_id: 'not-ora-canvas' }));
      var threw = false;
      try { await F.read(bad); } catch (e) { threw = /format_id/.test(e.message); }
      record('canvas-file-format #9: format_id mismatch rejected', threw, '');
    } catch (e) {
      record('canvas-file-format #9', false, 'threw: ' + (e.stack || e.message));
    }

    // 10. Malformed JSON rejected.
    try {
      var bad = new TextEncoder().encode('not json at all');
      var threw = false;
      try { await F.read(bad); } catch (e) { threw = /invalid JSON/.test(e.message); }
      record('canvas-file-format #10: malformed JSON rejected', threw, '');
    } catch (e) {
      record('canvas-file-format #10', false, 'threw: ' + (e.stack || e.message));
    }

    // 11. validate() accepts fixture; rejects image without image_data.
    try {
      var ok1 = F.validate(fixture);
      var bad = JSON.parse(JSON.stringify(fixture));
      bad.objects.push({ id: 'broken', kind: 'image', layer: 'background' });
      var ok2 = F.validate(bad);
      record('canvas-file-format #11: validate accepts fixture and rejects bad image',
        ok1.valid === true && ok2.valid === false,
        'valid_fixture=' + ok1.valid + ' bad_image=' + ok2.valid);
    } catch (e) {
      record('canvas-file-format #11', false, 'threw: ' + (e.stack || e.message));
    }

    // 12. isCompressed detection.
    try {
      var u = await F.write(fixture, { compressed: false });
      var c = await F.write(fixture, { compressed: true });
      record('canvas-file-format #12: isCompressed detects gzip magic',
        F.isCompressed(c) === true && F.isCompressed(u) === false,
        '');
    } catch (e) {
      record('canvas-file-format #12', false, 'threw: ' + (e.stack || e.message));
    }
  }
};
